### PR TITLE
Metrics - Take into account http request to time-serie-server while observing result save

### PR DESCRIPTION
**Pull request by thangqp**
This PR is following to the discussion on the PR of Franck: https://github.com/gridsuite/dynamic-simulation-server/pull/82

### CODE REVIEW

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java [hunk 1]
```diff
@@ -176,7 +176,7 @@ public Mono<DynamicSimulationResult> updateResult(UUID resultUuid, DynamicSimula
         Objects.requireNonNull(resultUuid);
         List<TimeSeries> timeSeries = new ArrayList<>(result.getCurves().values());
         StringTimeSeries timeLine = result.getTimeLine();
-        return Mono.zip(
+        return dynamicSimulationObserver.observe("results.save", runContext, () -> Mono.zip(
```
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java:179]: This obesravtion oberves only the time of the mono creation !

**Review by FranckLecuyer — COMMENTED** [review]
```
@Bean
    public Consumer<Message<byte[]>> consumeRun() {
        return message -> {
            LOGGER_BROKER_INPUT.debug("consumeRun {}", message);
            DynamicSimulationResultContext resultContext = DynamicSimulationResultContext.fromMessage(message);
            try {
                DynamicSimulationResult dynamicSimulationResult = dynamicSimulationObserver.observeRun("run", resultContext.getRunContext(), () -> Objects.requireNonNull(run(resultContext.getRunContext()).block()));
                updateResult(resultContext.getResultUuid(), resultContext.getRunContext(), dynamicSimulationResult);
                Message<String> sendMessage = MessageBuilder
                    .withPayload("")
                    .setHeader(DynamicSimulationResultContext.HEADER_RESULT_UUID, resultContext.getResultUuid().toString())
                    .setHeader(DynamicSimulationResultContext.HEADER_RECEIVER, resultContext.getRunContext().getReceiver())
                    .build();
                notificationService.emitResultDynamicSimulationMessage(sendMessage);
                LOGGER.info("Dynamic simulation complete (resultUuid='{}')", resultContext.getResultUuid());
            } catch (Exception e) {
                if (!(e instanceof CancellationException)) {
                    LOGGER.error(FAIL_MESSAGE, e);
                    // send fail notification
                    Message<String> sendMessage = new DynamicSimulationFailedContext(resultContext.getRunContext().getReceiver(),
                            resultContext.getResultUuid(),
                            FAIL_MESSAGE + " : " + e.getMessage(),
                            resultContext.getRunContext().getUserId()
                            ).toMessage();

                    notificationService.emitFailDynamicSimulationMessage(sendMessage);
                    // delete result entity in server's db
                    dynamicSimulationWorkerUpdateResult.deleteResult(resultContext.getResultUuid());
                }
            }
        };
    }

    public void updateResult(UUID resultUuid, DynamicSimulationRunContext runContext, DynamicSimulationResult result) {
        Objects.requireNonNull(resultUuid);
        List<TimeSeries> timeSeries = new ArrayList<>(result.getCurves().values());
        StringTimeSeries timeLine = result.getTimeLine();
        dynamicSimulationObserver.observe("results.save", runContext, () -> Objects.requireNonNull(Mono.zip(
                    timeSeriesClient.sendTimeSeries(timeSeries).subscribeOn(Schedulers.boundedElastic()),
                    timeSeriesClient.sendTimeSeries(Arrays.asList(timeLine)).subscribeOn(Schedulers.boundedElastic())
                )
                .map(uuidTuple -> {
                    UUID timeSeriesUuid = uuidTuple.getT1().getId();
                    UUID timeLineUuid = uuidTuple.getT2().getId();
                    DynamicSimulationStatus status = result.isOk() ? DynamicSimulationStatus.CONVERGED : DynamicSimulationStatus.DIVERGED;

                    dynamicSimulationWorkerUpdateResult.doUpdateResult(resultUuid, timeSeriesUuid, timeLineUuid, status);
                    return result;
                }).block()));
    }

```

File: src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java:179]: @Bean
    public Consumer<Message<byte[]>> consumeRun() {
        return message -> {
            LOGGER_BROKER_INPUT.debug("consumeRun {}", message);
            DynamicSimulationResultContext resultContext = DynamicSimulationResultContext.fromMessage(message);
            try {
                DynamicSimulationResult dynamicSimulationResult = dynamicSimulationObserver.observeRun("run", resultContext.getRunContext(), () -> Objects.requireNonNull(run(resultContext.getRunContext()).block()));
                updateResult(resultContext.getResultUuid(), resultContext.getRunContext(), dynamicSimulationResult);
                Message<String> sendMessage = MessageBuilder
                    .withPayload("")
                    .setHeader(DynamicSimulationResultContext.HEADER_RESULT_UUID, resultContext.getResultUuid().toString())
                    .setHeader(DynamicSimulationResultContext.HEADER_RECEIVER, resultContext.getRunContext().getReceiver())
                    .build();
                notificationService.emitResultDynamicSimulationMessage(sendMessage);
                LOGGER.info("Dynamic simulation complete (resultUuid='{}')", resultContext.getResultUuid());
            } catch (Exception e) {
                if (!(e instanceof CancellationException)) {
                    LOGGER.error(FAIL_MESSAGE, e);
                    // send fail notification
                    Message<String> sendMessage = new DynamicSimulationFailedContext(resultContext.getRunContext().getReceiver(),
                            resultContext.getResultUuid(),
                            FAIL_MESSAGE + " : " + e.getMessage(),
                            resultContext.getRunContext().getUserId()
                            ).toMessage();

                    notificationService.emitFailDynamicSimulationMessage(sendMessage);
                    // delete result entity in server's db
                    dynamicSimulationWorkerUpdateResult.deleteResult(resultContext.getResultUuid());
                }
            }
        };
    }

    public void updateResult(UUID resultUuid, DynamicSimulationRunContext runContext, DynamicSimulationResult result) {
        Objects.requireNonNull(resultUuid);
        List<TimeSeries> timeSeries = new ArrayList<>(result.getCurves().values());
        StringTimeSeries timeLine = result.getTimeLine();
        dynamicSimulationObserver.observe("results.save", runContext, () -> Objects.requireNonNull(Mono.zip(
                    timeSeriesClient.sendTimeSeries(timeSeries).subscribeOn(Schedulers.boundedElastic()),
                    timeSeriesClient.sendTimeSeries(Arrays.asList(timeLine)).subscribeOn(Schedulers.boundedElastic())
                )
                .map(uuidTuple -> {
                    UUID timeSeriesUuid = uuidTuple.getT1().getId();
                    UUID timeLineUuid = uuidTuple.getT2().getId();
                    DynamicSimulationStatus status = result.isOk() ? DynamicSimulationStatus.CONVERGED : DynamicSimulationStatus.DIVERGED;

                    dynamicSimulationWorkerUpdateResult.doUpdateResult(resultUuid, timeSeriesUuid, timeLineUuid, status);
                    return result;
                }).block()));
    }

[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java:179]: This should work, no ?

[the remark this answers is not in the document]
Comment (thangqp) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java]: DONE

[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java]: To suppress warning, you cat encapsulate the call of Mono.zip in updateResult with Objects.requireNonNull,
like it is done in consumeRun method

**Review by FranckLecuyer — APPROVED** [review]

File: src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (thangqp) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java]: You mean a warning in IDEA by SonaLint?

[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java]: yes

[the remark this answers is not in the document]
Comment (thangqp) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java]: See with Franck, is not Sonar warning, it is one from IDEA, so we ignore..

[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java]: I think all warnings should not be ignored (Inetlleij, Sonar, ..)

File: src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java [hunk 2]
```diff
@@ -143,17 +143,23 @@ public Consumer<Message<byte[]>> consumeRun() {
...
-                        notificationService.emitResultDynamicSimulationMessage(sendMessage);
-                        LOGGER.info("Dynamic simulation complete (resultUuid='{}')", resultContext.getResultUuid());
-                    }).block();
+                // run simulation
+                DynamicSimulationResult dynamicSimulationResult = dynamicSimulationObserver.observeRun("run",
+                        resultContext.getRunContext(),
+                        () -> Objects.requireNonNull(run(resultContext.getRunContext()).block()));
+
+                // save result
+                updateResult(resultContext.getResultUuid(), resultContext.getRunContext(), dynamicSimulationResult);
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java:152]: Move the observation here ?

Comment (thangqp) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java]: DONE

File: src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (thangqp) [reply src/main/java/org/gridsuite/ds/server/service/DynamicSimulationWorkerService.java]: This is a warning on IDEA of Franck but not in mine, anyway with commit https://github.com/gridsuite/dynamic-simulation-server/pull/83/commits/f7a6f99bdf47d782a58d5a147abb86f18c5e74d6, we have no longer this problem..

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**